### DISCUSSION TITLE
build a heterogenous graph dataset from a pandas df

### DISCUSSION THREAD

**Discussion by vats2004** [pyg-team/pytorch_geometric#9028] [Q&A]
I am very new to PyG. I am trying to build a heterogenous PyG dataset from a pandas df. **How does one assign label while building this hetero dataset?** One of the nodes types I have is "Player" (with attributes age, id, name, level). I want the level ('professional' or 'junior') to be the label. At this point, I am just want to learn how to build the right PyG dataset for a given data in json/csv/df. 

Goal: given a player-node, I want to classify if the player is a 'professional' or a 'junior' level player. => based on player's attributes, the org he/she belongs to, the events he had played at etc.

```
    def build_dataset(self, edge_index, org_X, player_X, event_X, labels_y):
        data = HeteroData()
        data['org'].x = org_X
        data['player'].x = player_X
        data['event'].x = event_X

        data['event', 'is_related_to', 'event'].edge_index = edge_index
        data['player'].y = labels_y

        return data
```

This is the output I get:

```
Input df:
  event_id                   event_type  org_id                        org_name org_location    player_id player_name  age related_event_id  player_level
0    1-ab3                   qualifiers     305               milan tennis club        Milan  1-b7a3-52d2        Alex   16   [4-ab3, 3-ab3]        Junior
1    2-ab3                    Wimbledon      76            Nadal tennis academy       madrid  2-b7a3-52d2         Bob   24          [1-ab3]  Professional
2    3-ab3     womens tennis qualifiers     185                Griz tennis club     budapest  3-b7a3-52d2        Mary   17          [4-ab3]        Junior
3    4-ab3  US professional tennis club     285  Nick Bolletieri Tennis Academy        tampa  4-b7a3-52d2         Joe   17   [1-ab3, 3-ab3]        Junior
4    5-ab3        womens tennis circuit     305               milan tennis club        Milan  5-b7a3-52d2        Bolt   17          [4-ab3]        Junior

Sorted and converted to numeric: input df:
  related_event_id  org_id  org_name  org_location  player_id  player_name  player_level  event_id  event_type  age
1              [0]       0         1             2          1            1             1         1           1    2
2              [3]       1         0             1          2            4             0         2           4    1
3           [0, 2]       2         2             3          3            3             0         3           0    1
0           [3, 2]       3         3             0          0            0             0         0           2    0
4              [3]       3         3             0          4            2             0         4           3    1

Org X:
tensor([[1, 2],
        [0, 1],
        [2, 3],
        [3, 0],
        [3, 0]], dtype=torch.int32)

Player X:
tensor([[1, 2],
        [4, 1],
        [3, 1],
        [0, 0],
        [2, 1]], dtype=torch.int32)

Event X:
tensor([[1],
        [4],
        [0],
        [2],
        [3]], dtype=torch.int32)

Labels:
[1 0 0 0 0]

Edge Index:
tensor([[1, 2, 3, 3, 0, 0, 4],
        [0, 3, 0, 2, 3, 2, 3]])

Graph Data:
HeteroData(
  org={ x=[5, 2] },
  player={
    x=[5, 2],
    y=[5],
  },
  event={ x=[5, 1] },
  (event, is_related_to, event)={ edge_index=[2, 7] }
)


Number of graphs: 3
Number of nodes: 15
Number of edges: 7
Number of node-features: {'org': 2, 'player': 2, 'event': 1}
Number of edge-features: {('event', 'is_related_to', 'event'): 0}
Edges are directed: True
Graph has isolated nodes: True
Graph has loops: False
Node Types: ['org', 'player', 'event']
Edge Attributes: 15

Data Metadata:
(['org', 'player', 'event'], [('event', 'is_related_to', 'event')])
/Users/vats/.pyenv/versions/3.10.13/lib/python3.10/site-packages/torch_geometric/data/data.py:1188: UserWarning: The node types {'org', 'player'} are isolated and are not referenced by any edge type 
  warnings.warn(msg)
```

**Other related questions:**

1. Why the number of graphs is 3? I expected it to be only one.
2. The number of nodes is 15. I expected it to be just 5.
3. What should I do to address the warning msg about {org, player} being isolated? Please advise. This warning is coming from `validate()`.

Any help is very much appreciated. I have a small dataset with 5 nodes and associated code. Will be happy to provide.

**Comment by wsad1** [comment]
> Why the number of graphs is 3? I expected it to be only one.

How did you compute `num_graphs` . As in what function did you use to compute. Is this the number of node types?

> The number of nodes is 15. I expected it to be just 5.

Each node type has 5 nodes and there are 3 node types 'org', 'player' and `event'.

> What should I do to address the warning msg about {org, player} being isolated? Please advise. This warning is coming from validate().

The only edge is `'event', 'is_related_to', 'event'`. Add edges to 'org` and 'player` to avoid this message.

**Reply by vats2004** [reply]
Thanks for the answers @wsad1 . I used:
```
        print(f'Number of graphs: {len(graph_data)}')
        print(f'Number of nodes: {graph_data.num_nodes}')
        print(f'Number of edges: {graph_data.num_edges}')
```
In total, in my input data, I have only 5 rows...where each row has org, player and event related information. I thought having 3 entities (node types) would be a reasonable way to begin. And, that is what I did ...have three node-types and added only their features to their respective x.

**Reply by rusty1s** [reply]
`len(HeteroData)` would return the number of different attributes in your data (similar to `len(dict)`), which is 3 in your case (`x`, `y`, `edge_index`). Each node type represents a set of nodes, so PyG expects to operate on 15 nodes in your case. If you just want to add different set of features to your nodes, you should concatenate in the feature dimension:
```python
x = torch.cat([org_X, player_X, event_X], dim=1)
